### ISSUE TITLE
InMemoryExporter for Metric is buggy

### ISSUE THREAD

**Issue by cijothomas** [open-telemetry/opentelemetry-dotnet#2361]
Unlike Traces, the Batch<Metric> provided to the exporter is re-used by SDK to keep metrics in memory. This means that, after the Export(Batch<Metric>), returns, the memory must not be read by the exporters. If they do so, they might be reading incorrect values. 

InMemoryExporter is buggy in this way - the fix would be for the InMemoryExporter to "deep copy" the Metric, MetricPoints, before Export() call returns, so that it has own copy, unaffected by further aggregations.

**Comment by TimothyMothra** [comment]
@cijothomas, I'm interested in working on this. 
I stumbled into a repro of this issue.

**Comment by cijothomas** [comment]
@utpilla might know more but i think we tried to fix this.. but hit some blockers.

**Comment by reyang** [comment]
Here are the principles I would suggest:
1. Perf - it shouldn't affect the performance of the main story (e.g. anyone who is not using InMemoryExporter will not see perf drop).
2. Simplicity - folks who are using InMemoryExporter should not do extra things. No extra public APIs introduced.

**Comment by TimothyMothra** [comment]
Questions:
- Are we committed to deep-copying Metric?
- Can we change the behavior of InMemoryExporter for Metrics only?




### Regarding deep-copy Metric

The original suggestion was to deep-copy the `Metric`, but there are challenges to this approach:
1. Cannot use MemoryStream/Binaryormatter because classes are not marked [Serializeable].
2. Cannot use XmlSerializer or JsonSerializer because classes do not have parameterless constructors.
3. A Reflection-based deep-copy might be possible but will have issues w/ parameterless constructor. This could be worked around, but this would be expensive to maintain because any change in any class could affect this.
4. An approach using dotnet's [MemberwiseClone](https://docs.microsoft.com/en-us/dotnet/api/system.object.memberwiseclone?view=net-6.0#examples) pattern might work. 
   ```
   MemberwiseClone Metric
   MemberwiseClone Metric.AggregatorStore
   Deep-Copy Metric.AggregatorStore.MetricPoint[]
   ```
    This would require some new public methods.
    Currently, both the AggregatorStore and MetricPoint array are readonly fields.
    Either the readonly would need to be removed or would need to use reflection to override the fields.


### Can we change the behavior of InMemoryExporter?

Currently the InMemoryExporter is initialized with a collection of exported items and the exporter adds to that collection. We could side-step this issue by clearing the collection everytime we begin a Metric export.

I opened a Draft PR to illustrate how this would work #2907

**Comment by TimothyMothra** [comment]
## Update
Discussed this issue in the 2/22/2022 meeting.

- Will clear collection in exporter.
- Will update [doc](https://github.com/open-telemetry/opentelemetry-dotnet/tree/main/docs/metrics/extending-the-sdk) with description of change in behavior. Users will be expected to perform any operations with exported items before exporting again. - Will work on this statement in the PR.
- Will update existing unit tests that are manually clearing.


At this time, we will not implement deep-copy. Nothing prevents us from providing this in the future if needed.

**Comment by reyang** [comment]
> * Will clear collection in exporter.
> * Will update [doc](https://github.com/open-telemetry/opentelemetry-dotnet/tree/main/docs/metrics/extending-the-sdk) with description of change in behavior. Users will be expected to perform any operations with exported items before exporting again. - Will work on this statement in the PR.
> * Will update existing unit tests that are manually clearing.

Would this proposal work? I guess no:

```csharp
counter.Add(...);
provider.ForceFlush();
counter.Add(...); // would this affect what has already been exported (since the point is reused)?
```

**Comment by TimothyMothra** [comment]
> // would this affect what has already been exported (since the point is reused)?

No, working with the `Counter` doesn't affect the `Metric` nor `MetricPoint` until calling `ForceFlush`. 



The weirdness happens after the second flush and the side effect depends on what a user does with the exported items.

### Metric
In this example, the exported collection has a reference to the same `Metric` and will always have updated values after the second flush.
```
counter.Add(...);
meterProvider.ForceFlush();
var metric1 = exportedItems[0];

counter.Add(...);
meterProvider.ForceFlush();
var metric2 = exportedItems[0];

Assert.Same(metric1, metric2); // User grabbed a reference to the Metric before the second flush. This will have updated values.
```

### MetricPoint
In this example, metricPoint1 and metricPoint2 are copies of structs and their snapshotValues will be unique.
```
counter.Add(...);
meterProvider.ForceFlush();
var metricPointsEnumerator1 = exportedItems[0].GetMetricPoints().GetEnumerator(); 
metricPointsEnumerator1.MoveNext(); 
var metricPoint1 = metricPointsEnumerator1.Current; 

counter.Add(...);
meterProvider.ForceFlush();
var metricPointsEnumerator2 = exportedItems[0].GetMetricPoints().GetEnumerator(); 
metricPointsEnumerator2.MoveNext(); 
var metricPoint2 = metricPointsEnumerator2.Current; 
```



### MetricPoint & HistogramBuckets
Contrary to the previous example, when working with Histograms the `MetricPoint` can no longer be trusted to have unique values.
The `MetricPoint` struct has a reference to `HistogramBuckets` and will always reflect the updated value.
```
var histogram = meter.CreateHistogram<int>("histogram");

histogram.Record(...);
meterProvider.ForceFlush();

var metricPointsEnumerator1 = exportedItems[0].GetMetricPoints().GetEnumerator(); 
metricPointsEnumerator1.MoveNext(); 
var metricPoint1 = metricPointsEnumerator1.Current; 

metricPoint1.GetHistogramCount(); // This will reflect the first Record()
metricPoint1.GetHistogramSum(); // this will reflect the first Record()

histogram.Record(...);
meterProvider.ForceFlush();

metricPoint1.GetHistogramCount(); // This will reflect the updated value
metricPoint1.GetHistogramSum(); // this will reflect the updated value
```

**Comment by TimothyMothra** [comment]
Per conversation in https://github.com/open-telemetry/opentelemetry-dotnet/pull/3071#discussion_r834603581

- I'm going to use the doc PR (#3071) to update the extending-the-sdk guide.
- will focus on an approach to clone the exported metrics. We want this change in 1.2

@cijothomas, can you assign this item to the 1.2 milestone for tracking?

**Comment by cijothomas** [comment]
Sure. The approach shown here (https://github.com/open-telemetry/opentelemetry-dotnet/pull/3071#discussion_r833602498), while "cheating" (😄 ), serves the purpose and may be followed.

**Comment by cijothomas** [comment]
To avoid any more delays to 1.2.0 stable release, removing this item from 1.2.0 milestone. This means InMemoryExporter will not be doing 1.2.0 stable release, along with SDK. (similar to Prometheus Exporter).
Will address this for 1.3.0 stable (dates/features for that will be updated in milestones page soon)

**Comment by stevejgordon** [comment]
@martinjt This one looks to have been resolved in #3266 so I think it can be closed.
